packages/runtime-core/__tests__/rendererPortal.spec.ts

```diff
@@ -1,3 +1,130 @@
+import {
+  resetOps,
+  mockWarn,
+  h,
+  Portal,
+  render,
+  nodeOps,
+  serializeInner,
+  nextTick,
+  ref
+} from '@vue/runtime-test'
+
 describe('renderer: portal', () => {
-  test.todo('should work')
+  mockWarn()
+
+  test('portals require a target', async () => {
+    // TOODO: Remove this. Implementation detail of querySelector in tests
+    resetOps()
+
+    const Parent = () => h(Portal, [h(Child)])
+
+    const Child = () => h('div', 'child')
+
+    const Comp = () => [h(Parent), h('div', { id: 'target' })]
+
+    const root = nodeOps.createElement('div')
+    render(h(Comp), root)
+
+    expect('Invalid Portal target on mount:').toHaveBeenWarned()
+
+    // expect(serializeInner(root)).toBe(
+    //   `<div id="target"></div>`
+    // )
+  })
+
+  test('portals require a target in DOM', async () => {
+    // TOODO: Remove this. Implementation detail of querySelector in tests
+    resetOps()
+
+    const Parent = () => h(Portal, { target: '#target1' }, [h(Child)])
+
+    const Child = () => h('div', 'child')
+
+    const Comp = () => [h(Parent), h('div', { id: 'target2' })]
+
+    const root = nodeOps.createElement('div')
+    render(h(Comp), root)
+
+    expect('Invalid Portal target on mount:').toHaveBeenWarned()
```

Comment:
I think that it would be useful to have the error message display the target _selector_ since the target will always be `null` when this error is shown.